### PR TITLE
Use a new yaml_merge.py tool to properly generate the expanded api.yaml file

### PR DESCRIPTION
**Pull request by schenedx**
## Description

With this change, we can use the new yml_merge.py utility to expand the api.yaml. The tool would leverage both the pyyaml and ruamel.ymal "merge-expand" option to get the job done. However, the output won't have any comments because pyyaml strip out comments in api.yaml.

@edx/masters-neem Please review

### CODE REVIEW

File: yml_merge.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+"""
+This utility would invoke pyyaml to load and dump the DRY yaml file
+with both merge key anchors and shared dictionary elements. 
+Pyyaml would handle the merge key elements properly with overrides.
+However, Pyyaml would not expand anchored shared dictionary elements.
+We then use the ruamel commandline tool "merge-expan" to make sure shared
```
Comment (zacharis278) [inline yml_merge.py]: nit. typo on "merge-expand"

**Review by zacharis278 — APPROVED** [review]
💯

File: Makefile [hunk 2]
```diff
@@ -112,7 +112,7 @@ detect_changed_source_translations: ## check if translation files are up-to-date
 validate_translations: fake_translations detect_changed_source_translations ## install fake translations and check if translation files are up-to-date
 
 api_generated: ## generates an expanded verison of api.yaml for consuming tools that cannot read yaml anchors
-	yaml merge-expand api.yaml .api-generated.yaml
+	python yml_merge.py api.yaml .api-generated.yaml
 
 validate_api_committed: ## check to make sure any api.yaml changes have been committed to the expanded document
 	bash -c "diff .api-generated.yaml <(yaml merge-expand api.yaml -)"
```
Comment (zacharis278) [inline Makefile]: `validate_api_committed` needs to use your new task as well

Comment (schenedx) [reply Makefile]: Yes. Found that out. Could you check if what I am doing is still good?

**Review by kdmccormick — COMMENTED** [review]
Let's keep the root directory clean. Please move `yml_merge.py` to `scripts/`.

Also, to make sure that the scripts in `scripts/` get their code style checked, can you change in the Makefile:

```Makefile
quality:
    pycodestyle registrar *.py
    ...
```
to:
```Makefile
quality:
    pycodestyle registrar scripts
    ...
```

Other than that, looks good!

File: yml_merge.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+"""
```
Comment (iloveagent57) [inline yml_merge.py]: We have a `scripts/` directory, can this live in there?

**Comment by kdmccormick** [comment]
One more nit: `yaml_merge.py` would probably be a more appropriate name

File: yml_merge.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        output_path = '/tmp/generated.yml'
+
+    with open(output_path, 'w') as wp:
+        yaml.dump(data, wp)
+
+    # Now, use ruamel yaml to merge-expand
+    os.system('yaml merge-expand {0} {1}'.format(output_path, sys.argv[2]))
+
+    if output_path != sys.argv[2]:
+        os.remove(output_path)
```
Comment (iloveagent57) [inline yml_merge.py]: Is it safe to remove this file if the `os.system(...)` call fails?

File: scripts/yml_merge.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+"""
+This utility would invoke pyyaml to load and dump the DRY yaml file
```
Comment (iloveagent57) [inline scripts/yml_merge.py]: what's a "DRY" yaml file?

File: yml_merge.py (hunk 3 above)
Comment (schenedx) [reply yml_merge.py]: Done

File: yml_merge.py (hunk 4 above)
Comment (schenedx) [reply yml_merge.py]: Yes in the grand scheme of things. The worst would be deleting the `.api-generated.yaml`, which we have in github so we don't necessarily lose it. 
All other problems, we can easily notice and fix locally.

File: scripts/yml_merge.py (hunk 5 above)
Comment (schenedx) [reply scripts/yml_merge.py]: I made it up. It's the file with anchors and alias we need this process for. What's your name for it?

Comment (kdmccormick) [reply scripts/yml_merge.py]: "Don't Repeat Yourself"?

**Review by kdmccormick — APPROVED** [review]
Send it :+1:

**Pull request merged**